waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,475 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/[options, sugar, sequtils],
+  stew/[results, byteutils],
+  chronicles,
+  chronos,
+  metrics,
+  libp2p/protocols/protocol,
+  libp2p/stream/connection,
+  libp2p/crypto/crypto,
+  eth/p2p/discoveryv5/enr
+import
+  ../common/nimchronos,
+  ../common/enr,
+  ../waku_core,
+  ../waku_archive,
+  ../waku_store/[client, common],
+  ../waku_enr,
+  ../node/peer_manager/peer_manager,
+  ./raw_bindings,
+  ./common,
+  ./session
+
+logScope:
+  topics = "waku sync"
+
+type WakuSync* = ref object of LPProtocol
+  storage: NegentropyStorage
+  peerManager: PeerManager
+  maxFrameSize: int
+
+  syncInterval: timer.Duration # Time between each syncronisation attempt
+  relayJitter: Duration # Time delay until all messages are mostly received network wide
+  transferCallBack: Option[TransferCallback] # Callback for message transfer.
```

Comment:
ATM it is only set IF we have access to `wakuArchive` and `wakusStoreClient`.

We could change it so that the setup fail instead?

Reply:
> We could change it so that the setup fail instead?

I think it makes sense to fail as soon as possible and completely stop the `wakunode2` at the beginning if we consider a wrong initial setup.

> edit: we need to be able to not set the callbacks for testing purposes.

Maybe we can use `isNil()` instead? Not a blocking deal though :)